### ISSUE TITLE
Prompt script block System.NullReferenceException

### ISSUE THREAD

**Issue by annahosanna** [dahlbyk/posh-git#602]
- posh-git version/path: 0.7.3.1  C:\tools\poshgit\dahlbyk-posh-git-9bda399\src
- PowerShell version: 4.0
- git version 2.18.0.windows.1
- OS: Microsoft Windows NT 6.1.7601 Service Pack 1

Posh-git installed via Chocolatey

After posh-git install Powershell prompt appeared as:
PS>

It seems that there was an issue setting currentPath within the prompt script block:
$currentPath = $ExecutionContext.SessionState.InvokeCommand.ExpandString($GitPromptSettings.DefaultPromptPath);
>>
Object reference not set to an instance of an object.
    + CategoryInfo          : OperationStopped: (:) [], NullReferenceException
    + FullyQualifiedErrorId : System.NullReferenceException

However DefaultPromptPath seems to be defined:
write-host $global:GitPromptSettings.DefaultPromptPath
$(Get-PromptPath)

To resolve the issue I was able to redefine the prompt function with an updated prompt script block in the profile, after importing the posh-git module, by updating this single line:
$currentPath = $(Get-PromptPath)

Is there a better way to resolve this?

Thank you

**Comment by dahlbyk** [comment]
Are you able to get a full stack trace? Try setting `$GitPromptSettings.Debug = $true`.

**Comment by dahlbyk** [comment]
If you restore posh-git back to the original code, do you get the same error if you set `$GitPromptSettings.DefaultPromptPath = '$pwd'`?

**Comment by annahosanna** [comment]
Updating GitPrompt.ps1 to set GitPromptSettings.Debug to $true had no effect (prompt was still "PS>" )

**Comment by annahosanna** [comment]
Your second suggestion worked:

PS>cd .\myrepo
PS>$GitPromptSettings.DefaultPromptPath = '$pwd'
C:\Users\me\Documents\GitHub\myrepo [master ≡]>

**Comment by dahlbyk** [comment]
Is calling `Get-PromptPath` directly successful as well? I'm struggling to reconcile that `'$pwd'` can be expanded but `'$(Get-PromptPath)'` can't be.

**Comment by annahosanna** [comment]
PS>$(Get-PromptPath)
C:\Users\me\Documents\GitHub\myrepo
PS>
PS>$GitPromptSettings.DefaultPromptPath = $(Get-PromptPath)
C:\Users\me\Documents\GitHub\myrepo [master ≡]>

**Comment by annahosanna** [comment]
Seems like it is using the literal value "$(Get-PromptPath)" rather than the result of from evaluating that function.

PS>write-host $Global:GitPromptSettings.DefaultPromptPath
$(Get-PromptPath)

PS>$GitPromptSettings.DefaultPromptPath = $(Get-PromptPath)
C:\Users\me\Documents\GitHub\myrepo [master ≡]> write-host $Global:GitPromptSettings.DefaultPromptPath
C:\Users\me\Documents\GitHub\myrepo

**Comment by annahosanna** [comment]
I think it is from GitPrompt.ps1. DefaultPromptPath should be using quotes not ticks.

DefaultPromptPath                           = '$(Get-PromptPath)'
should be:
DefaultPromptPath                           = "$(Get-PromptPath)"

**Comment by rkeithhill** [comment]
Indeed, you have to use single quotes when setting the value.  I think we should update that to make it also handle a scriptblock. This would be less susceptible to error:
```
$GitPromptSettings.DefaultPromptPath = { Get-PromptPath }
```

**Comment by dahlbyk** [comment]
Handling a ScriptBlock would be neat.

I still don't know what the nullref is from, though.

**Comment by annahosanna** [comment]
I just added:

$Global:GitPromptSettings.DefaultPromptPath = $(Get-PromptPath)

to my profile to work around the issue. As far as I am concerned you can close this issue.

**Comment by annahosanna** [comment]
This does not work for me either:

$Global:GitPromptSettings.DefaultPromptPath='$($ExecutionContext.SessionState.Path.CurrentLocation.Path)'

**Comment by annahosanna** [comment]
Nor this:

```
C:\Users\me\Documents\GitHub\myrepo [master ≡]> $ExecutionContext.SessionState.InvokeCommand.ExpandString('$($ExecutionContext.SessionState.Path.CurrentLocation.Path)')
Exception calling "ExpandString" with "1" argument(s): "Object reference not set to aninstance of an object."
At line:1 char:1
+ $ExecutionContext.SessionState.InvokeCommand.ExpandString('$($ExecutionContext.S ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [], MethodInvocationException
    + FullyQualifiedErrorId : NullReferenceException
```

**Comment by annahosanna** [comment]
But this works:


`C:\Users\me\Documents\GitHub\myrepo [master ≡]> $ExecutionContext.SessionState.InvokeCommand.InvokeScript($ExecutionContext.SessionState.InvokeCommand.NewScriptBlock('$($ExecutionContext.SessionState.Path.CurrentLocation.Path)'))`

**Comment by annahosanna** [comment]
This works too:

```
PS>$Global:GitPromptSettings.DefaultPromptPath='$(Get-PromptPath)'
PS>$ExecutionContext.SessionState.InvokeCommand.InvokeScript($ExecutionContext.SessionState.InvokeCommand.NewScriptBlock($Global:GitPromptSettings.DefaultPromptPath))
C:\Users\me\Documents\GitHub\myrepo
```

**Comment by rkeithhill** [comment]
This:
```
$Global:GitPromptSettings.DefaultPromptPath='$(Get-PromptPath)'
```
Is what we expect folks to use and is how it is set by default - see https://github.com/dahlbyk/posh-git/blob/5dcdb9df9ffd9a10c65dd02c75352fee979ae019/src/GitPrompt.ps1#L112

The single quotes defers interpolation until we call `ExpandString` in the posh-git provided prompt function - see https://github.com/dahlbyk/posh-git/blob/5dcdb9df9ffd9a10c65dd02c75352fee979ae019/src/posh-git.psm1#L55

So, when you have the property set like this do you still the get NulRefEx?

**Comment by annahosanna** [comment]
Yes.  It works with single quotes, and performs dynamic expansion. It creates a script block from the text in single quotes, and then evaluates the script block.

**Comment by rkeithhill** [comment]
> It creates a script block from the text in single quotes

That's not correct.  By using single quotes, you're telling PowerShell not to do interpolation.  When you use double quotes PowerShell interpolates first and then assigns to the setting e.g.:
```
$Global:GitPromptSettings.DefaultPromptPath='$(Get-PromptPath)'
$Global:GitPromptSettings.DefaultPromptPath
'$(Get-PromptPath)'

$Global:GitPromptSettings.DefaultPromptPath="$(Get-PromptPath)"
$Global:GitPromptSettings.DefaultPromptPath
C:\User\hillr
```
The point of calling ` $ExecutionContext.SessionState.InvokeCommand.ExpandString()` is that it will do the interpolation for us at the point in *time we call it*. 

Try this:
```
$foo = '$(2+2)'
$foo
$(2+2)

$ExecutionContext.SessionState.InvokeCommand.ExpandString($foo)
4
```

But let me ask this again, when you have the property set like this (in single quotes) do you still the get NulRefEx?

**Comment by dahlbyk** [comment]
> This works in the GitPromptScritpBlock
> replacing:
> 
> $currentPath = $ExecutionContext.SessionState.InvokeCommand.ExpandString($GitPromptSettings.DefaultPromptPath)
> With:
> 
> $currentPath = $ExecutionContext.SessionState.InvokeCommand.InvokeScript($ExecutionContext.SessionState.InvokeCommand.NewScriptBlock($GitPromptSettings.DefaultPromptPath))

So the hypothesis is that some part of `$ExecutionContext.SessionState.Path.CurrentLocation` is returning `$null`, but only when called via `ExpandString()`? You can use `$ExecutionContext.SessionState.InvokeCommand.ExpandString('$ExecutionContext.SessionState.Path.CurrentLocation')` to narrow down what part is problematic. Does `$ExecutionContext.SessionState.InvokeCommand.ExpandString('$(Get-Location)')` work?

https://github.com/dahlbyk/posh-git/blob/5dcdb9df9ffd9a10c65dd02c75352fee979ae019/src/Utils.ps1#L282-L287

**Comment by rkeithhill** [comment]
FYI, you need to put that expression (that accesses properties) inside PowerShell sub-expression syntax e.g.:
```
$ExecutionContext.SessionState.InvokeCommand.ExpandString('$($ExecutionContext.SessionState.Path.CurrentLocation)')
```

**Comment by annahosanna** [comment]
@rkeithhill Sorry I must have intercommunicated something. 
1)  I no longer get the nullref using $ExecutionContext.SessionState.InvokeCommand.InvokeScript instead of $ExecutionContext.SessionState.InvokeCommand.ExpandString.
2) I understood that you were using `'` to control when interpolation was done.

**Comment by annahosanna** [comment]
@dahlbyk 
No

```
PS>$ExecutionContext.SessionState.InvokeCommand.ExpandString('$(Get-Location)')
Object reference not set to an instance of an object.
    + CategoryInfo          : OperationStopped: (:) [], NullReferenceException
    + FullyQualifiedErrorId : System.NullReferenceException
```

**Comment by annahosanna** [comment]
@dahlbyk 
```
PS>$ExecutionContext.SessionState.InvokeCommand.ExpandString('$ExecutionContext.SessionState.Path.CurrentLocation')
System.Management.Automation.EngineIntrinsics.SessionState.Path.CurrentLocation
```

**Comment by dahlbyk** [comment]
@annahosanna sorry, try this:

```
$ExecutionContext.SessionState.InvokeCommand.ExpandString('$($ExecutionContext.SessionState.Path.CurrentLocation)')
```

Or even this?

```
$ExecutionContext.SessionState.InvokeCommand.ExpandString('$pwd')
```

**Comment by annahosanna** [comment]
```
PS>$ExecutionContext.SessionState.InvokeCommand.ExpandString('$($ExecutionContext.SessionState.Path.CurrentLocation)')
Exception calling "ExpandString" with "1" argument(s): "Object reference not set to an
instance of an object."
At line:1 char:1
+ $ExecutionContext.SessionState.InvokeCommand.ExpandString('$($ExecutionContext.S ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [], MethodInvocationException
    + FullyQualifiedErrorId : NullReferenceException
```

**Comment by dahlbyk** [comment]
```
$ExecutionContext.SessionState.InvokeCommand.ExpandString('$($ExecutionContext.SessionState)')
```
or
```
$ExecutionContext.SessionState.InvokeCommand.ExpandString('$pwd')
```
❓

**Comment by annahosanna** [comment]
```
PS>$ExecutionContext.SessionState.InvokeCommand.ExpandString('$pwd')
C:\Users\me\Documents\GitHub
```

**Comment by annahosanna** [comment]
```
PS>$ExecutionContext.SessionState.InvokeCommand.ExpandString('$($ExecutionContext.SessionState)')
Exception calling "ExpandString" with "1" argument(s): "Object reference not set to an
instance of an object."
At line:1 char:1
+ $ExecutionContext.SessionState.InvokeCommand.ExpandString('$($ExecutionContext.S ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [], MethodInvocationException
    + FullyQualifiedErrorId : NullReferenceException
```

**Comment by annahosanna** [comment]
@rkeithhill , @dahlbyk 

I'm happy to answer questions; however, this issue does not need to remain open since I can use the defaults with the solution I mentioned earlier.
 
```
$currentPath = $ExecutionContext.SessionState.InvokeCommand.InvokeScript($ExecutionContext.SessionState.InvokeCommand.NewScriptBlock($GitPromptSettings.DefaultPromptPath))
```

**Comment by rkeithhill** [comment]
No but I would like to spin up a PS v4 VM to see if I can replicate this (later tonight).  BTW we can catch the scriptblock that gets created.  It doesn't need to be recreated each time.

**Comment by annahosanna** [comment]
Googling around this may be specific to certain versions of Windows (I'm using Windows 7, but I've also seen mention of this with 2012) with PS 4.

**Comment by dahlbyk** [comment]
The other workaround would be to add:

```powershell
if (-not $pathInfo) {
  $pathInfo = $pwd
}
```

**Comment by rkeithhill** [comment]
I was able to easily repro this on a Windows 7 VM updated to PS 4.  It seems the parser chokes on **any sub-expression** passed to `ExpandString()`:

![image](https://user-images.githubusercontent.com/5177512/43359742-fa80cbc8-9264-11e8-883a-b0f200dd0271.png)

Below you can see the full stack trace of where it chokes in our prompt function.  Looks like it chokes reading a `Line` property on some `InternalScriptPosition` object.

![image](https://user-images.githubusercontent.com/5177512/43359715-3e4e2324-9264-11e8-88d2-7d912a24637c.png)

Unfortunately the suggestion solution doesn't always work:

![image](https://user-images.githubusercontent.com/5177512/43359630-9fd13cc8-9262-11e8-8654-93a72b2b60c8.png)

But it does work with the **default case** and can be made to work in the scenario above with an extra set of double quotes e.g.:
```
$GitPromptSettings.DefaultPromptPath.Text = '"MyPath is $(Get-PromptPath)"'
```
However, this change could break folks on PS >= 5 who have customized their `DefaultPromptPath` setting e.g.:

```
07-28 12:39:55 11> $GitPromptSettings.DefaultPromptPath.Text = 'MyPath is $(Get-PromptPath)'
07-28 12:41:16 12> $ExecutionContext.SessionState.InvokeCommand.InvokeScript($ExecutionContext.SessionState.InvokeCommand.NewScriptBlock($GitPromptSettings.DefaultPromptPath.Text))
Exception calling "InvokeScript" with "1" argument(s): "The term 'MyPath' is not recognized as the name of a cmdlet,
function, script file, or operable program. Check the spelling of the name, or if a path was included, verify that the
path is correct and try again."
At line:1 char:1
+ $ExecutionContext.SessionState.InvokeCommand.InvokeScript($ExecutionC ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [], MethodInvocationException
    + FullyQualifiedErrorId : CommandNotFoundException
```

So I'm thinking we should conditionally apply this fix to PS <= v4.  Thoughts?  I suppose we should also verify the "fix" works on PS v3.

**Comment by dahlbyk** [comment]
How about a string wrapper inside `NewScriptBlock`? Roughly (on phone, untested):

```
$ExecutionContext.SessionState.InvokeCommand.InvokeScript(
  $ExecutionContext.SessionState.InvokeCommand.NewScriptBlock(
  "`"`$($($GitPromptSettings.DefaultPromptPath))`""
  ))
```

(Note this bug is for v0, so we don't want `.Text`)

**Comment by rkeithhill** [comment]
> (Note this bug is for v0, so we don't want .Text)

Got that.  But on my dev system (running V5 / V6.1) it is easier to test $ExecutionContext with my currently loaded posh-git v1 beta2.  Ultimately, the *text* passed into NewScriptBlock is the same. 

The string wrapper, as-is, still fails:

![image](https://user-images.githubusercontent.com/5177512/43361490-d1d6bb06-928c-11e8-93a7-978809b19cb3.png)
